genomeview/bamtrack.py

```diff
@@ -744,179 +747,172 @@ def layout(self, scale):
-                    j += 1
-                    k += 1
+            if is_fwd:
+                start_pos = read.reference_start
+                if start_pos < scale.start:
+                    bin_index = 0
+                else:
+                    bin_index = ((start_pos - scale.start) // self.bin_size) + 1
+            else:
+                end_pos = read.reference_end
+                if end_pos > scale.end:
+                    bin_index = 0
+                else:
+                    bin_index = ((scale.end - end_pos) // self.bin_size) + 1
+
+            for j in aligned_pos:
+                if scale.start <= j < scale.end:
+                    coverage[bin_index][j - scale.start] += 1
+
+        self._add_multi_coverage(scale, coverage)
+
+    def add_peak_coverage(self, scale):
+        read_ends = collections.Counter()
+
+        # flag to indicate which side RT started from
+        is_fwd = (
+            (scale.strand == "+" and self.priming_orientation == "5p")
+            or (scale.strand == "-" and self.priming_orientation == "3p")
+        )
+
+        for read in self._get_reads(scale):
+            if is_fwd:
+                read_ends[read.reference_start] += 1
+            else:
+                read_ends[read.reference_end] += 1
+
+        most_common = read_ends.most_common()
+        peaks = []
+        for i, (v, w) in enumerate(most_common[:BAMCoverageTrack.MAX_BINS]):
+            if min((abs(v - v2) for v2, _ in most_common[:i]), default=self.min_dist) >= self.min_dist:
+                peaks.append(v)
 
-                # Update the current layer with the cumulated depth
-                layers[bin_index] = (x_combined, y_updated)
+        coverage = collections.defaultdict(collections.Counter)
+
+        for read in self._get_reads(scale):
+            # get all the reference coordinates that are aligned to the read
+            aligned_pos = read.get_reference_positions()
+
+            # this shouldn't happen, should it? -> should not because fetch is used on the region so only reads aligning there will be returned
+            # could also check if any coordinates are in the region of interest
+            # if not aligned_pos:
+            #    continue
+
+            if is_fwd:
+                pos = read.reference_start
+            else:
+                pos = read.reference_end
+
+            bin_index = min(peaks, key=lambda p: abs(pos - p))
+            if abs(pos - bin_index) > self.min_dist:
+                bin_index = -1
+
+            for j in aligned_pos:
+                if scale.start <= j < scale.end:
+                    coverage[bin_index][j - scale.start] += 1
+
+        self._add_multi_coverage(scale, coverage)
+
+    def add_tagged_coverage(self, scale):
+        coverage = collections.defaultdict(collections.Counter)
+
+        for read in self._get_reads(scale):
+            if not read.has_tag(self.tag):
+                continue
```

Comment:
Should reads without the tag be grouped in an extra bin, optionally?

Reply:
In my use cases I don't think that'd be useful, but maybe in other situations.

Reply:
Made a change so that they are grouped in a separate secondary bin that gets added over all other layers and uses a different color palette.